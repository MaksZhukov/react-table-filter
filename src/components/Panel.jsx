import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Draggable from 'react-draggable';
import { Portal } from 'react-portal';
import FilterContainer from '../containers/Filters';
import Tables from '../containers/Tables';
import { OFFSETS_FILTERS } from '../constants';

class Panel extends PureComponent {
  startDragFilter = (event) => {
    document.querySelectorAll('.filters').forEach((filter) => {
      if (filter !== event.currentTarget) {
        filter.classList.remove('has-draggable');
      }
    });
    event.stopPropagation();
  }

  handleClickToggleFilters = () => {
    const { toggleFilters, id } = this.props;
    toggleFilters(id);
  }

  handleClickRemovePanel = () => {
    const { removePanel, id } = this.props;
    removePanel(id);
  }

  render() {
    const {
      isOpenFilters, id, x, y,
    } = this.props;

    return (
      <>
        <div className="panel">
          <button type="button" className="btn btn-filter" onClick={this.handleClickToggleFilters}>
            filter
          </button>
          <button type="button" className="btn" onClick={this.handleClickRemovePanel}>
            x
          </button>
          <span>
            â„–
            {' '}
            {id}
          </span>
          <Tables id={id} />
          {isOpenFilters
            && (
              <Portal node={document && document.getElementById('portal-filters')}>
                <Draggable
                  defaultClassName={classNames('react-draggable', 'has-draggable')}
                  onStart={this.startDragFilter}
                  defaultPosition={{
                    x: x * OFFSETS_FILTERS.width + OFFSETS_FILTERS.indentX,
                    y: y * OFFSETS_FILTERS.height + OFFSETS_FILTERS.indentY
                    + +(y > 0 ? OFFSETS_FILTERS.additionalIndentY : 0),
                  }}
                >
                  <div className="filters">
                    <FilterContainer id={id} />
                  </div>
                </Draggable>
              </Portal>
            )
          }
        </div>
      </>
    );
  }
}

Panel.propTypes = {
  id: PropTypes.number.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  toggleFilters: PropTypes.func.isRequired,
  isOpenFilters: PropTypes.bool.isRequired,
  removePanel: PropTypes.func.isRequired,
};

export default Panel;
