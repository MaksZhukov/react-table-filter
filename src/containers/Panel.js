import { connect } from 'react-redux';
import Panel from '../components/Panel';
import { toggleFilters, removePanel } from '../actions';


const mapStateToProps = (state, { id }) => ({
  isOpenFilters: state.get('panels').find(panel => panel.get('id') === id).get('isOpenFilters'),
});

const mapDispatchToProps = dispatch => ({
  toggleFilters: id => dispatch(toggleFilters(id)),
  removePanel: id => dispatch(removePanel(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Panel);
