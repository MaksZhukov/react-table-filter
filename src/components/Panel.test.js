import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import Panel from './Panel';


configure({ adapter: new Adapter() });
describe('component Panel', () => {
  const props = {
    id: 0,
    x: 0,
    y: 0,
    toggleFilters: jest.fn(),
    isOpenFilters: false,
    removePanel: jest.fn(),
  };
  it('render component Panel without crashing', () => {
    const wrapper = shallow(<Panel {...props} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('component Panel with properly props', () => {
    const wrapper = shallow(<Panel {...props} />);
    expect(wrapper.instance().props.id).toEqual(0);
    expect(wrapper.instance().props.x).toEqual(0);
    expect(wrapper.instance().props.y).toEqual(0);
    expect(typeof wrapper.instance().props.toggleFilters).toBe('function');
  });
  it('component Panel without state', () => {
    const wrapper = shallow(<Panel {...props} />);
    expect(wrapper.instance().state).toBe(null);
  });
  it('component Panel contain wrong number', () => {
    const wrapper = shallow(<Panel {...props} />);
    expect(wrapper.contains(
      <span>
        â„–
        {' '}
        1
      </span>,
    )).toBe(false);
  });
  it('component Panel has one btn open filter', () => {
    const wrapper = shallow(<Panel {...props} />);
    expect(wrapper.find('.btn-filter').length).toBe(1);
  });
});
