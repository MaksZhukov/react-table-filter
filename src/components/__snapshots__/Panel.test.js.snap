// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Panel render component Panel without crashing 1`] = `
<Fragment>
  <div
    className="panel"
  >
    <button
      className="btn btn-filter"
      onClick={[Function]}
      type="button"
    >
      filter
    </button>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <span>
      â„–
       
      0
    </span>
    <Connect(Tables)
      id={0}
    />
  </div>
</Fragment>
`;
